src/hotspot/share/utilities/globalDefinitions.hpp

```diff
@@ -1260,6 +1260,24 @@ JAVA_INTEGER_SHIFT_OP(>>, java_shift_right_unsigned, jlong, julong)
 
 #undef JAVA_INTEGER_SHIFT_OP
 
+// Some convenient bit shift operations that accepts a BasicType as the last
+// argument. These avoid potential mistakes with overloaded functions only
+// distinguished by lhs argument type.
+#define JAVA_INTEGER_SHIFT_BASIC_TYPE(FUNC)            \
+inline jlong FUNC(jlong lhs, jint rhs, BasicType bt) { \
+  if (bt == T_INT) {                                   \
+    return FUNC((jint) lhs, rhs);                      \
+  }                                                    \
+  assert(bt == T_LONG, "unsupported basic type");      \
+  return FUNC(lhs, rhs);                              \
+}
+
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_left)
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_right)
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_right_unsigned)
+
```

Comment:
Couldn't this be written as a template, so we could use java_shift_left<T_INT>(x,y) or java_shift_left<jint>(x.y)?

Reply:
Explicit `BasicType` as a function (instead of template) parameter allows more flexibility at runtime when `bt` is a variable. For example:

https://github.com/openjdk/jdk/blob/e9b42dcce268f32bd2ec3c01ac6221073b888538/src/hotspot/share/opto/addnode.cpp#L471